DigitalLearningSolutions.Data.Migrations/202204071530_PopulateUsersTableFromAccountsTables.cs

```diff
@@ -0,0 +1,415 @@
+
+            foreach (var emailGroup in delegateAccountsGroupedByEmail)
+            {
+                if (!string.IsNullOrWhiteSpace(emailGroup.Key))
+                {
+                    var delegatesToAttemptToLink = new List<DelegateAccount>();
+                    var delegateAccountsGroupedByCentre = emailGroup.GroupBy(da => da.CentreId);
+
+                    foreach (var centreGroup in delegateAccountsGroupedByCentre)
+                    {
+                        delegatesToAttemptToLink.Add(centreGroup.First());
+
+                        // All duplicate emails at a centre need new User accounts.
+                        // If there are more than one delegate accounts with the same email at a centre
+                        // we only want to keep the email on the first. The rest get reset to having a User with a guid email
+                        var othersAtCentre = centreGroup.Skip(1);
+
+                        foreach (var delegateAccount in othersAtCentre)
+                        {
+                            // Insert a new User with Guid email
+                            InsertNewUserForDelegateAccount(connection, delegateAccount, true);
+                        }
+                    }
+
+                    var allJobGroupsMatch =
+                        delegatesToAttemptToLink.Select(da => da.JobGroupId_deprecated).Distinct().Count() < 2;
+                    foreach (var delegateAccount in delegatesToAttemptToLink)
+                    {
+                        // Check for existing user with email
+                        var existingUserWithEmail = connection.QuerySingleOrDefault<User>(
+                            "SELECT * FROM Users WHERE PrimaryEmail = @email",
+                            new { email = emailGroup.Key }
+                        );
+
+                        if (existingUserWithEmail != null)
+                        {
+                            // If we find a user, we update any default data on that user
+                            UpdateExistingUserWithDelegateDetails(
+                                connection,
+                                existingUserWithEmail,
+                                delegateAccount,
+                                allJobGroupsMatch
+                            );
+                        }
+                        else
+                        {
+                            // Otherwise we create a new user with the email address
+                            InsertNewUserForDelegateAccount(connection, delegateAccount, false);
+                        }
+                    }
+                }
+                else
+                {
+                    // If we don't have a valid email we create new users for each with a guid email
+                    foreach (var delegateAccount in emailGroup)
+                    {
+                        InsertNewUserForDelegateAccount(connection, delegateAccount, true);

```

Comment:
See https://github.com/TechnologyEnhancedLearning/DLSV2/pull/1065#discussion_r850381154, same deal. If there is a single user we set this to now, if there are more users it can become unset.

Reply:
I mean that, in this particular case, we are inserting a new User for this particular DelegateAccount with a GUID email. We aren't going to be linking any other accounts to this, so we won't get a chance to update the DetailsLastChecked (and the DelegateAccounts equivalent), which will be set to the current date time.

So with that in mind, a populated DetailsLastChecked would imply that a user does not need to check their details, because we assume they are correct after this migration. But in this case, the primary email will be a GUID, which I think we want them to check and correct. Whenever the user logs in to an account with a GUID as the primary email, we plan to redirect them to the Edit Detail page so they can change it.

I think this also applies to the other scenario where we set the primary email to a GUID, but this was easier to explain in this circumstance.

Reply:
Ah OK, yeah I think you're right in the GUID case